src/components/views/dialogs/SlidingSyncOptionsDialog.tsx

```diff
@@ -0,0 +1,134 @@
+ */
+async function proxyHealthCheck(endpoint: string, hsUrl?: string): Promise<void> {
+    const controller = new AbortController();
+    const id = setTimeout(() => controller.abort(), 10 * 1000); // 10s
+    const res = await fetch(endpoint + "/client/server.json", {
+        signal: controller.signal,
+    });
+    clearTimeout(id);
+    if (res.status != 200) {
+        throw new Error(`proxyHealthCheck: proxy server returned HTTP ${res.status}`);
+    }
+    const body = await res.json();
+    if (body.server !== hsUrl) {
+        throw new Error(`proxyHealthCheck: client using ${hsUrl} but server is as ${body.server}`);
+    }
+    logger.info("sliding sync proxy is OK");
+}
+
+export const SlidingSyncOptionsDialog: React.FC<IDialogProps> = ({ onFinished }) => {
+    const cli = MatrixClientPeg.get();
+    const enabled = SettingsStore.getValue("feature_sliding_sync");
+    const currentProxy = SettingsStore.getValue("feature_sliding_sync_proxy_url");
+    const hasNativeSupport = useAsyncMemo(() => syncHealthCheck(cli).then(() => true, () => false), [], null);
+
+    let nativeSupport: string;
+    if (hasNativeSupport === null) {
+        nativeSupport = _t("Checking...");
+    } else {
+        nativeSupport = hasNativeSupport
+            ? _t("Your server has native support")
+            : _t("Your server lacks native support");
+    }
+
+    const validProxy = withValidation<undefined, { error?: Error }>({
+        async deriveData({ value }): Promise<{ error?: Error }> {
+            try {
+                await proxyHealthCheck(value, MatrixClientPeg.get().baseUrl);
+                return {};
+            } catch (error) {
+                return { error };
+            }
+        },
+        rules: [
+            {
+                key: "required",
+                test: async ({ value }) => !!value || hasNativeSupport,
+                invalid: () => _t("Your server lacks native support, you must specify a proxy"),
+            }, {
+                key: "working",
+                final: true,
+                test: async (_, { error }) => !error,
+                valid: () => _t("Looks good"),
+                invalid: ({ error }) => error?.message,
+            },
+        ],
+    });
+
+    return <TextInputDialog
+        title={_t("Sliding Sync configuration")}
+        description={nativeSupport}
+        placeholder={hasNativeSupport ? _t('Proxy URL') : _t('Proxy URL (optional)')}
+        value={currentProxy}
+        button={_t("Save")}
+        cancelButton={_t("Disable")}
+        validator={validProxy}
+        onFinished={(enable: boolean, proxyUrl: string) => {
+            if (enable) {
+                SettingsStore.setValue("feature_sliding_sync_proxy_url", null, SettingLevel.DEVICE, proxyUrl);
+                onFinished(true);
+                if (enabled && proxyUrl !== currentProxy) {
+                    // We're changing proxy URL here, cause a reload
+                    PlatformPeg.get().reload();
```

Comment:
If we turn off SS mode we need to logout the user, else we can end up breaking to-device messages and therefore E2EE which will result in UISIs. This is due to:
 - proxy has A,B,C and has its own since token system to deliver them to the client
 - before they can be delivered the user turns off SS mode
 - EW syncs normally but A,B,C will never be delivered as the proxy ACKed them previously
 
 The only safe way to handle this is to logout the user and make them login again, which will then start delivering to-device messages from scratch for the new device.

Reply:
We could ditch the `Disable` button and add copy which says "To disable SS you must log out and back in" - which would clear localStorage and thus disable all the things

Reply:
we should probably warn before enabling too

Reply:
![image](https://user-images.githubusercontent.com/2403652/186877016-92347768-eb8f-40e7-93bd-6a172fbcf344.png)
![image](https://user-images.githubusercontent.com/2403652/186877219-19fa522a-da1d-496d-bf5e-df8a63a9682f.png)
![image](https://user-images.githubusercontent.com/2403652/186877269-5ac54dc4-6184-47ee-accb-08aa241e1890.png)